TrackletAlgorithm/MatchEngineUnit.h

```diff
@@ -0,0 +1,349 @@
+#ifndef MATCHENGINEUNIT_H
+#define MATCHENGINEUNIT_H
+
+#include "Constants.h"
+#include "CandidateMatchMemory.h"
+#include "TrackletProjectionMemory.h"
+#include "VMStubMEMemoryCM.h"
+#include "VMProjectionMemory.h"
+#include "ProjectionRouterBuffer.h"
+#include "AllStubMemory.h"
+#include "AllProjectionMemory.h"
+#include "FullMatchMemory.h"
+#include "hls_math.h"
+#include <iostream>
+#include <fstream>
+#include <bitset>
+#include "MatchProcessor_parameters.h"
+
+template<int nbits>
+static const ap_uint<1 << nbits> hasOneStub() {
+  ap_uint<1 << nbits> onlyOneStub(0);
+  onlyOneStub[1] = 1;
+  return onlyOneStub;
+}
+
+template<int VMProjType> class MatchEngineUnitBase {};
+
+template<>
+class MatchEngineUnitBase<BARREL> {
+ public:
+  enum BitWidths{
+    kNBitsBuffer=3,
+    kNMatchEngines=4
```

Comment:
Why is there another kNMatchEngines with the same name as the one in Constants.h?

Reply:
Removed.